cpp/daal/src/algorithms/dtrees/gbt/treeshap.h

```diff
@@ -0,0 +1,476 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/**
+ * Original TreeSHAP algorithm by Scott Lundberg, 2018
+ * https://arxiv.org/abs/1802.03888
+ *
+ * Fast TreeSHAP algorithm v1 and v2 by Jilei Yang, 2021
+ * https://arxiv.org/abs/2109.09847.
+ */
+
+/*
+//++
+//  Implementation of the treeShap algorithm
+//--
+*/
+
+#ifndef __TREESHAP_H__
+#define __TREESHAP_H__
+
+#include "services/daal_defines.h"
+#include "services/error_handling.h"
+#include "src/algorithms/dtrees/dtrees_feature_type_helper.h"
+#include "src/algorithms/dtrees/gbt/gbt_model_impl.h"
+#include "src/services/service_arrays.h"
+#include "src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i"
+
+namespace daal
+{
+namespace algorithms
+{
+namespace gbt
+{
+namespace treeshap
+{
+using gbt::internal::FeatureIndexType;
+using gbt::internal::ModelFPType;
+using FeatureTypes = algorithms::dtrees::internal::FeatureTypes;
+
+/**
+ * Determine the requested version of the TreeSHAP algorithm set in the
+ * environment variable SHAP_VERSION.
+ * Returns fallback if SHAP_VERSION is not set.
+*/
+uint8_t getRequestedAlgorithmVersion(uint8_t fallback);
+
+/**
+ * Decision Path context
+*/
+struct PathElement
+{
+    int featureIndex                 = 0;
+    float zeroFraction               = 0;
+    float oneFraction                = 0;
+    float partialWeight              = 0;
+    PathElement()                    = default;
+    PathElement(const PathElement &) = default;
+};
+
+namespace internal
+{
+
+namespace v0
+{
+
+void extendPath(PathElement * uniquePath, size_t uniqueDepth, float zeroFraction, float oneFraction, int featureIndex);
+void unwindPath(PathElement * uniquePath, size_t uniqueDepth, size_t pathIndex);
+float unwoundPathSum(const PathElement * uniquePath, size_t uniqueDepth, size_t pathIndex);
+
+/** Recursive treeShap function
+ * \param nodeIndex the index of the current node in the tree, counted from 1
+ * \param depth how deep are we in the tree
+ * \param uniqueDepth how many unique features are above the current node in the tree
+ * \param parentUniquePath a vector of statistics about our current path through the tree
+ * \param parentZeroFraction what fraction of the parent path weight is coming as 0 (integrated)
+ * \param parentOneFraction what fraction of the parent path weight is coming as 1 (fixed)
+ * \param parentFeatureIndex what feature the parent node used to split
+ * \param conditionFraction what fraction of the current weight matches our conditioning feature
+ */
+template <typename algorithmFPType, bool hasUnorderedFeatures, bool hasAnyMissing>
+inline void treeShap(const gbt::internal::GbtDecisionTree * tree, const algorithmFPType * x, algorithmFPType * phi,
```

Comment:
Do we want to return status?

Reply:
I return a status for all functions that allocate memory. The recursive parts of `treeShap` are kept `void`.